cameronmaddern/pollish_app
# PR #5 User can vote on a poll

Adds the ability for a signed-in user to vote on a poll, delete this vote and change it to another option.

**To test:**
- Check out the PR branch
- Run `cdk deploy BackendStack-test-<username> --profile <your-profile>`
- Run `scripts/fetch_graphql.sh`
- This will update the schema to allow `DELETE` methods on vote objects

Currently, the percentages are pulled on the first load. If the user votes, then the vote tallies are updated locally _only_.

**TODO in future:**
- Re-fetch poll votes periodically so that the percentages seen by the end user remain fresh.
- Currently, the option state is updated optimistically, that is before the fetch to the API is made. This is fine since we expect these requests to typically succeed, however methods should be added to ensure that in the unlikely event of a failed call, the state is updated as needed.
  - Note that I have experimented with making the API call first, _then_ updating the state however the slight delay in the `await` makes for a significantly worse experience
- Streamline/refactor the voting function. Currently it is duplicated across both `text_poll` and `image_poll` components. See comments elsewhere regarding whether these components can be merged.
